DCP2 Terra exporter

rolando-ebi opened the pull request:
* Added ExperimentGraph to represent an experiment graph in a structured and type-safe format
 * Added GraphCrawler to support new links.json structure and implement a refactored/cleaned-up experiment graph crawling algorithm
 * Added DcpStagingClient, for writing metadata to the DCP2 staging bucket and downloading data files from S3 where necessary
* Added TerraExporter for generating and exporting ExperimentGraphs to DCP2 ,given assay/analysis process UUIDs

rolando-ebi committed:
refactored exporter: separated graph generation from data export

rolando-ebi committed:
simplified graph crawling in lieu of new links.json

rolando-ebi committed:
bug fixes

rolando-ebi committed:
implemented Terra exporter, with dummy GCS file staging

rolando-ebi committed:
implemented the DcpStagingClient (renamed from GcsStagingClient)

rolando-ebi committed:
updated test ingest resources to include the 'dcpVersion' and 'type' fields

rolando-ebi committed:
testing the graph crawler against the 'in-memory' ingest-api

rolando-ebi committed:
moved graph crawl helper methods to MetadataService

rolando-ebi committed:
implemented simple updates

aaclan-ebi commented on exporter/graph/graph_crawler.py:
```diff
@@ -0,0 +1,55 @@
+from exporter.graph.experiment_graph import ExperimentGraph, Link, Input, Output, ProtocolLink
+from typing import List, Iterable, Dict
+from functools import reduce
+from _operator import iconcat
```
might be better to import from operator instead _operator.

aaclan-ebi commented on exporter/terra/dcp_staging_client.py:
```diff
@@ -0,0 +1,118 @@
+    def write_metadata(self, metadata: MetadataResource):
+        dest_object_key = f'metadata/{metadata.concrete_type()}/{metadata.uuid}_{metadata.dcp_version}.json'
+
+        if metadata.metadata_type == "file":
```
I think the condition here should be != "file" .

aaclan-ebi commented on exporter/terra/dcp_staging_client.py:
```diff
@@ -0,0 +1,118 @@
+        return StringIO(json.dumps(d))
+
+    @staticmethod
+    def convert_file_metadata(file_metadata: MetadataResource) -> MetadataResource:
+        """
+        This function is file-schema aware, and must transform the filename property to
+        the form of {dataFileUuid}_{dataFileVersion}_{filename}.
+
+        The dataFileVersion isn't defined in ingest, so we'll re-use the dcpVersion
+        of the file_metadata here
+        :return: DCPv2 MVP-ready file metadata
+        """
+        if file_metadata.full_resource is not None:
+            data_file_uuid = file_metadata.full_resource["dataFileUuid"]
+            data_file_version = file_metadata.dcp_version
+            filename = file_metadata.full_resource["content"]["file_core"]["file_name"]
+
+            new_file_metadata = deepcopy(file_metadata)
+            new_name = f'{data_file_uuid}_{data_file_version}_{filename}'
+            new_file_metadata.full_resource = file_metadata.full_resource
+            new_file_metadata.full_resource["content"]["file_core"]["file_name"] = new_name
+
+            return new_file_metadata
+        else:
```
Do we need to do the patching of file content-type + checksum fields here?

aaclan-ebi replied:
looks like it's not decided yet.

rolando-ebi replied:
We might - it hasn't been agreed yet

aaclan-ebi commented on exporter/terra/dcp_staging_client.py:
```diff
@@ -0,0 +1,118 @@
+
+        if metadata.metadata_type == "file":
+            data_stream = DcpStagingClient.dict_to_json_stream(metadata.get_content())
+            self.write_to_staging_bucket(dest_object_key, data_stream)
```
for files shared in multiple graphs like biomaterials and protocols, should we consider concurrency in staging the files?

rolando-ebi replied:
As in should we consider not uploading the same file twice? Yep, this should be done. We should check if a file exists before attempting to upload

rolando-ebi replied:
Need to add code to check if a file already exists at an object path, good catch

aaclan-ebi commented on exporter/graph/graph_crawler.py:
```diff
@@ -0,0 +1,55 @@
+    def experiment_graph_for_process(self, process: MetadataResource) -> ExperimentGraph:
+        return self._crawl(process)
+
+    def _crawl(self, process: MetadataResource) -> ExperimentGraph:
+        partial_graph = ExperimentGraph()
+
+        process_info = self.process_info(process)
+        partial_graph.nodes.add_nodes(process_info.inputs + process_info.outputs + process_info.protocols + [process])
+        partial_graph.links.add_link(GraphCrawler.link_for(process_info))
+
+        processes_to_crawl = GraphCrawler.flatten([self.metadata_service.get_derived_by_processes(i) for i in process_info.inputs])
+
+        return reduce(lambda g1, g2: g1.extend(g2),
+                      map(lambda proc: self._crawl(proc), processes_to_crawl),
+                      partial_graph)
```
very concise

rolando-ebi replied:
Cheers. We might be able to use this for driving the archival/manifest-generator process later on as well

aaclan-ebi commented on exporter/terra/dcp_staging_client.py:
```diff
@@ -0,0 +1,118 @@
+            data_stream = DcpStagingClient.dict_to_json_stream(metadata.get_content())
+            self.write_to_staging_bucket(dest_object_key, data_stream)
+        else:
+            converted_file_metadata = DcpStagingClient.convert_file_metadata(metadata)
```
in the old exporter logic, we're populating provenance fields, should it also be included here?

rolando-ebi replied:
Good point. I haven't seen anything explicit about it, but we probably will have to. Will confirm in the main design doc

aaclan-ebi commented on exporter/terra/dcp_staging_client.py:
```diff
@@ -0,0 +1,118 @@
+            self.write_to_staging_bucket(dest_object_key, data_stream)
+
+    def write_links(self, link_set: LinkSet, experiment_uuid: str, experiment_version: str):
+        dest_object_key = f'links/{experiment_uuid}_{experiment_version}.json'
```
shouldn't this be `links/{links_id}_{version}_{project_id}.json`

rolando-ebi replied:
As of today yep! Was just agreed to by Dan and Hannes. Good catch

aaclan-ebi commented on exporter/terra/terra_exporter.py:
```diff
@@ -0,0 +1,43 @@
+        self.graph_crawler = graph_crawler
+        self.dcp_staging_client = dcp_staging_client
+
+    def export(self, process_uuid, submission_uuid, experiment_uuid, experiment_version):
```
what's the experiment_uuid / version? is it the same as bundle uuid + version?

rolando-ebi replied:
Yep it will be the UUID+version included in the new-assay messages generated by core, currently named "bundleUuid" I believe. Since we're not using bundles anymore we should consider renaming these. Downstream developers like the term "experiment" or "experiment graph", I also think it's a good name for assays/analysis "bundles"

rolando-ebi committed:
added project_uuid to the links.json gcs object key/path

fixed boto3 s3 download streams to work as inputs to gcs upload streams

added provenance infomation to exported metadata, refactored MetadataResource

rolando-ebi committed:
fixed tests, removed StagingClient and related tests

rolando-ebi committed:
generating links.json with the help of new SchemaService

rolando-ebi committed:
graph crawler performs concurrent requests to ingest-api

rolando-ebi committed:
don't upload files if they already exist

rolando-ebi committed:
aside: async_receiver for handling messages asynchronously while responding to amqp heartbeats

rolando-ebi committed:
added TerraExportListener - listening for and handling new experiment and update messages

aaclan-ebi commented on exporter/terra/dcp_staging_client.py:
```diff
@@ -65,7 +66,8 @@ def write_data_file(self, data_file: DataFile):
         staging_bucket: storage.Bucket = self.gcs_client.bucket("test-ingest-terra", "terra-ingest")
         blob: storage.Blob = staging_bucket.blob(object_key)
-        blob.upload_from_file(data_stream)
+        if not blob.exists():
+            blob.upload_from_file(data_stream)
```
I think it would be nice to get the checksum and override it if checksum is different in case the second process trying to upload is trying to correct the first uploaded json. If it's not trivial to add that i think it's fine, I don't think it's needed for now.

rolando-ebi replied:
Yeah I considered that. The spec says that we mustn't upload a file if it already exists at the path. Maybe we could have a flag to force-override for when we have to correct something due to a bug

rolando-ebi committed:
setting up the terra export listener in __main__

rolando-ebi committed:
refactored terra AMQP

rolando-ebi committed:
configurable GCS bucket, prefix, project for DcpStagingClient

rolando-ebi committed:
unset user_project param when interfacing with the staging bucket

rolando-ebi committed:
terra listener: prefetch count 1, fixed expected message format

rolando-ebi committed:
extra logging for tracking terra export progress

rolando-ebi committed:
added file_integrity to exported File metadata

rolando-ebi committed:
supporting supplementary files in the experiment graph

rolando-ebi committed:
testing LinkSet

rolando-ebi committed:
fixed tests

rolando-ebi committed:
moved file_integrity fields into top level of file_core

rolando-ebi committed:
driving manifest generation with experiment graph + graph crawler

rolando-ebi commented:
@MightyAx, based on previous comments regarding sharing code for archiving and exporting, I made the following commit for the manifest generator component: https://github.com/ebi-ait/ingest-exporter/pull/2/commits/c9a8f4d806a166c37c928593a6d191a1b44b306c

aaclan-ebi commented:
Changes to the manifest receiver looks good! Thanks for making the changes.

rolando-ebi committed:
handling write concurrency
